### PR TITLE
[FEATURE] Create expect_column_values_to_be_secure_passwords.py

### PR DESCRIPTION
**Pull request by spencerhardwick**
This expectation checks if the columns contain secure passwords. 

In this experimental expectation, the tests run but they don't print. 

Contributors: @aworld1 @spencerhardwick @carolli014

### CODE REVIEW

**Comment by kyleaton** [comment]
@cla-bot check

**Comment by spbail** [comment]
#2404 also implements a password expectation. We should decide which one to use.

**Comment by spencerhardwick** [comment]
> #2404 also implements a password expectation. We should decide which one to use.

My biggest concern with our submission was modularity around different password use cases - ie some sites require special characters, some don't. Some sites have a minimum requirement, some don't, etc. Does it look like the other implementation addresses this problem better than ours? Is it worth merging solutions?

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
    condition_value_keys = ("min_length", "min_uppercase", "min_lowercase", "min_special", "min_digits",
                            "max_consec_numbers", "max_consec_letters")
```

Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: The above keys should appear in `condition_value_keys` since the keys are used in computing the metric value

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
        def _pandas(cls, column, min_length, min_uppercase, min_lowercase, min_special, min_digits, max_consec_numbers, max_consec_letters, **kwargs):
        def matches_password_requirements(x):
            x = str(x)
```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py:86]:
```suggestion


```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: I think you meant to put these parameters in the top-level method. Declaring the parameters in the `condition_value_keys` attribute above (see first comment) is required so the metrics machinery knows to pull those keys from the Expectation config kwargs and pass them in here.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: Please replace this docstring with one appropriate for the expectation.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
        "maturity": "experimental",
        "tags": ["experimental", "column map expectation"],
        "package": "experimental_expectations",
        "contributors": [
            "@spencerhardwick",
            "@aworld1",
            "@carolli014",
        ],
        "requirements": [],
```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
    success_keys = (
        "min_length",
        "min_uppercase",
        "min_lowercase",
        "min_special",
        "min_digits",
        "max_consec_numbers",
        "max_consec_letters",
        "mostly"
    )
```

Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: Since the success/failure of the expectation depends on these kwargs, they should be declared in the `success_keys` attribute.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
    default_kwarg_values = {
        "row_condition": None,
        "condition_parser": None,  # we expect this to be explicitly set whenever a row_condition is passed
        "mostly": 1,
        "result_format": "BASIC",
        "include_config": True,
        "catch_exceptions": True,
        "min_length": 8,
        "min_uppercase": 1,
        "min_lowercase": 1,
        "min_special": 1,
        "min_digits": 1,
        "max_consec_numbers": 99,
        "max_consec_letters": 99
    }
```

Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: Rather than hardcoding these default values in the metric, they can be defined here.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
    def _question_renderer(
        cls, configuration, result=None, language=None, runtime_configuration=None
    ):
        column = configuration.kwargs.get("column")
        #password = configuration.kwargs.get("password")
        mostly = "{:.2%}".format(float(configuration.kwargs.get("mostly", 1)))

        return f'Are at least {mostly} of all values in column "{column}" secure passwords?'
```

Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: Since this is a `ColumnMapExpectation`, it automatically gets an optional `mostly` parameter, so a user can specify a threshold (e.g. a `mostly` value of `0.5` means the expectation will pass if 50% of column values pass the expectation). Adjusted the renderer accordingly.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
    def _answer_renderer(
        cls, configuration=None, result=None, language=None, runtime_configuration=None
    ):
        column = result.expectation_config.kwargs.get("column")
        #password = result.expectation_config.kwargs.get("password")
        mostly = "{:.2%}".format(float(configuration.kwargs.get("mostly", 1)))
        
        if result.success:
            return f'At least {mostly} of all values in column "{column}" are secure passwords.'
        else:
            return f'Less than {mostly} of all values in column "{column}" are secure passwords.'

```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
    # @classmethod
    # @renderer(renderer_type="renderer.prescriptive")
    # @render_evaluation_parameter_string
    # def _prescriptive_renderer(
    #     cls,
    #     configuration=None,
    #     result=None,
    #     language=None,
    #     runtime_configuration=None,
    #     **kwargs,
    # ):
    #     runtime_configuration = runtime_configuration or {}
    #     include_column_name = runtime_configuration.get("include_column_name", True)
    #     include_column_name = (
    #         include_column_name if include_column_name is not None else True
    #     )
    #     styling = runtime_configuration.get("styling")
    #     params = substitute_none_for_missing(
    #         configuration.kwargs,
    #         ["column", "mostly", "row_condition", "condition_parser"],
    #     )
    #     return [
    #         RenderedStringTemplateContent(
    #             **{
    #                 "content_block_type": "string_template",
    #                 "string_template": {
    #                     "template": template_str,
    #                     "params": params,
    #                     "styling": styling,
    #                 },
    #             }
    #         )
    #     ]
```

Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: Since this renderer wasn't complete, it was causing tests to fail. I've commented it out for now since it's not strictly required to merge your PR. If you'd like it included in the PR, please fill out the renderer.

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
                        "unexpected_list": ["AAAAAAAA", "asdf5hu!", "a!s4D"],
```

Comment (roblim) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]: Unexpected values need to appear here otherwise the test will fail.

**Review by roblim — CHANGES_REQUESTED** [review]
Thanks for the submission @spencerhardwick - awesome work! I ran `run_diagnostics` on the expectation and was getting errors/failures - I added some suggested changes to address these. I also think my suggested changes address your previous comment about modularity - with my suggested changes, a user can now specify "min_length", "min_uppercase", "min_lowercase", "min_special", "min_digits", "max_consec_numbers", and "max_consec_letters" (i.e. `expect_column_values_to_be_secure_passwords(column="app_password", "min_length"=22, "min_special"=3)`). In future revisions, you might want to adjust the logic to allow users to pass in `None` for the various parameters (e.g. if there is no `min_lowercase` constraint, etc.).

If you're okay with my suggested changes, the only thing left to do before merging the PR is updating the docstring (it currently has the docstring for the regex expectation). Let me know if you have any questions! 

For reference, here's a link to the acceptance checklist we use, as well as other relevant docs:
- https://github.com/great-expectations/great_expectations/blob/develop/docs/contributing/contrib_pull_request_acceptance_checklist.rst
- https://github.com/great-expectations/great_expectations/tree/develop/contrib

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py
Comment (roblim) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py]:
```suggestion
class ExpectColumnValuesToBeSecurePasswords(ColumnMapExpectation):
    """Expect column entries to be secure passwords, as defined by expectation parameters.
    
    expect_column_values_to_be_secure_passwords is a \
    :func:`column_map_expectation <great_expectations.execution_engine.execution_engine.MetaExecutionEngine
    .column_map_expectation>`.

    Args:
       min_length (int): minimum length
       min_uppercase (int): minimum number of uppercase letters
       min_lowercase (int): minimum number of lowercase letters
       min_special (int): minimum number of special characters (not letters and not digits)
       min_digits (int): minimum numbers of digits
       max_consec_numbers (int): max number of consecutive numbers
       max_consec_letters (int): max number of consecutive letters

    Keyword Args:
        mostly (None or a float between 0 and 1): \
            Return `"success": True` if at least mostly fraction of values match the expectation. \
            For more detail, see :ref:`mostly`.

    Other Parameters:
        result_format (str or None): \
            Which output mode to use: `BOOLEAN_ONLY`, `BASIC`, `COMPLETE`, or `SUMMARY`.
            For more detail, see :ref:`result_format <result_format>`.
        include_config (boolean): \
            If True, then include the expectation config as part of the result object. \
            For more detail, see :ref:`include_config`.
        catch_exceptions (boolean or None): \
            If True, then catch exceptions and include them as part of the result object. \
            For more detail, see :ref:`catch_exceptions`.
        meta (dict or None): \
            A JSON-serializable dictionary (nesting allowed) that will be included in the output without \
            modification. For more detail, see :ref:`meta`.

    Returns:
        An ExpectationSuiteValidationResult

        Exact fields vary depending on the values passed to :ref:`result_format <result_format>` and
        :ref:`include_config`, :ref:`catch_exceptions`, and :ref:`meta`.
    """
```

**Review by roblim — CHANGES_REQUESTED** [review]

**Review by roblim — APPROVED** [review]

**Pull request merged**
